src/services/codefixes/importFixes.ts

```diff
@@ -561,12 +574,37 @@ namespace ts.codefix {
         return some(declarations, decl => !!(getMeaningFromDeclaration(decl) & meaning));
     }
 
-    export function forEachExternalModuleToImportFrom(checker: TypeChecker, from: SourceFile, allSourceFiles: ReadonlyArray<SourceFile>, cb: (module: Symbol) => void) {
+    export function forEachExternalModuleToImportFrom(checker: TypeChecker, host: LanguageServiceHost, from: SourceFile, allSourceFiles: ReadonlyArray<SourceFile>, cb: (module: Symbol) => void) {
+        const { allowsImporting } = createLazyPackageJsonDependencyReader(from, host);
         forEachExternalModule(checker, allSourceFiles, (module, sourceFile) => {
-            if (sourceFile === undefined || sourceFile !== from && isImportablePath(from.fileName, sourceFile.fileName)) {
+            if (sourceFile === undefined && allowsImporting(stripQuotes(module.getName()))) {
                 cb(module);
             }
+            else if (sourceFile && sourceFile !== from && isImportablePath(from.fileName, sourceFile.fileName)) {
+                const moduleSpecifier = getModuleSpecifierFromFileName(sourceFile.fileName);
+                if (!moduleSpecifier || allowsImporting(moduleSpecifier)) {
+                    cb(module);
+                }
+            }
         });
+
+        function getModuleSpecifierFromFileName(fileName: string): string | undefined {

```

Comment:
This function feels very unscientific but I couldn’t find an obvious better way. Did I overlook something?

Reply:
`src/compiler/moduleSpecifiers.ts` has some functions for finding specifiers that is used in declaration emit and afaik some other quickfixes.